benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/time_step_sycl.cpp

```diff
@@ -0,0 +1,686 @@
+#include <iostream>
+#include <chrono>
+#include <vector>
+#include <cstdlib>
+#include <CL/sycl.hpp>
+
+using namespace cl::sycl;
+
+// Uncomment line below to use TIMER
+// #define USE_TIMER
+
+#ifdef USE_TIMER
+#include "timing.h"
+#endif
+
+queue * workqueue;
+bool first_time = true;
+
+extern "C" void c_invoke_time_step(
+        // Fields
+        double * ssha_t,
+        double * sshn_t,
+        double * sshn_u,
+        double * sshn_v,
+        double * hu,
+        double * hv,
+        double * un,
+        double * vn,
+        double * ua,
+        double * ht,
+        double * ssha_u,
+        double * va,
+        double * ssha_v,
+        // Grid
+        int * tmask,
+        double * area_t,
+        double * area_u,
+        double * area_v,
+        double * dx_u,
+        double * dx_v,
+        double * dx_t,
+        double * dy_u,
+        double * dy_v,
+        double * dy_t,
+        double * gphiu,
+        double * gphiv,
+        // Device Pointers are passed as void* and the invoke will cast them
+        // to the desired type, and by reference as the value has to be stored.
+        void * &ssha_t_view_p,
+        void * &sshn_t_view_p,
+        void * &sshn_u_view_p,
+        void * &sshn_v_view_p,
+        void * &hu_view_p,
+        void * &hv_view_p,
+        void * &un_view_p,
+        void * &vn_view_p,
+        void * &ua_view_p,
+        void * &ht_view_p,
+        void * &ssha_u_view_p,
+        void * &va_view_p,
+        void * &ssha_v_view_p,
+        void * &tmask_view_p,
+        void * &area_t_view_p,
+        void * &area_u_view_p,
+        void * &area_v_view_p,
+        void * &dx_u_view_p,
+        void * &dx_v_view_p,
+        void * &dx_t_view_p,
+        void * &dy_u_view_p,
+        void * &dy_v_view_p,
+        void * &dy_t_view_p,
+        void * &gphiu_view_p,
+        void * &gphiv_view_p,
+        // Scalars
+        int istep,
+        int internal_xstart,
+        int internal_xstop,
+        int internal_ystart,
+        int internal_ystop,
+        int width,
+        int total_size,
+        double rdt,
+        double cbfr,
+        double visc,
+        double omega,
+        double d2r,
+        double g
+        ){
+
+#ifdef USE_TIMER
+    TimerInit();
+    TimerStart("Sycl SetUp");
+#endif
+
+    if(first_time){
+        std::cout << "List of SYCL devices:" << std::endl;
+        auto devices = device::get_devices();
+        int index = 0;
+        for(auto dev : devices){
+            auto platform_name = dev.get_platform().get_info<info::platform::name>();
+            auto dev_name = dev.get_info<info::device::name>();
+            std::cout <<  " " << index++ << " ) " << platform_name << " : " <<
+                dev_name << std::endl;
+        }
+
+        int selected_device = 0;
+        if(const char* env_p = std::getenv("SYCL_PLATFORM")){
+            selected_device = std::stoi(env_p);
```

Comment:
Please could you add a check here that the selected device is valid.

Reply:
I added some basic checks and exception catching around this line